File: java/lib/src/main/java/com/zourzouvillys/httpsig/Components.java
package com.zourzouvillys.httpsig;

import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Extracts component values from HTTP messages per RFC 9421 Section 2.
 *
 * Handles both derived components (@method, @target-uri, etc.) and
 * header fields (plain, ;sf, ;bs, ;key).
 */
final class Components {

    private Components() {}

    /**
     * Extract the value of a component from the message.
     *
     * @param cid    the component to extract
     * @param msg    the HTTP message (request or response)
     * @param reqMsg the related request message (for ;req on responses, or null)
     * @return the extracted string value
     */
    static String extract(ComponentIdentifier cid, HttpMessage msg, HttpMessage reqMsg)
            throws HttpSigException {

        // ;req means extract from the request instead
        if (cid.hasParam("req")) {
            if (reqMsg == null) {
                throw new HttpSigException(";req used but no request message provided");
            }
            // strip the ;req param and extract from the request
            var strippedParams = new java.util.LinkedHashMap<>(cid.params());
            strippedParams.remove("req");
            var strippedCid = ComponentIdentifier.withParams(cid.name(), strippedParams);
            return extract(strippedCid, reqMsg, null);
        }

        String name = cid.name();
        if (name.startsWith("@")) {
            return extractDerived(name, cid, msg);
        }
        return extractHeader(name, cid, msg);
    }

    // ---- Derived components (Section 2.2) ----

    private static String extractDerived(String name, ComponentIdentifier cid, HttpMessage msg)
            throws HttpSigException {
        return switch (name) {
            case "@method" -> {
                requireRequest(msg, name);
                yield msg.method().toUpperCase();
            }
            case "@target-uri" -> {
                requireRequest(msg, name);
                yield msg.url().toASCIIString();
            }
            case "@authority" -> {
                requireRequest(msg, name);
                yield extractAuthority(msg.url());
            }
            case "@scheme" -> {
                requireRequest(msg, name);
                yield msg.url().getScheme().toLowerCase();
            }
            case "@request-target" -> {
                requireRequest(msg, name);
                yield extractRequestTarget(msg.url());
            }
            case "@path" -> {
                requireRequest(msg, name);
                String path = msg.url().getRawPath();
                yield (path == null || path.isEmpty()) ? "/" : path;
            }
            case "@query" -> {
                requireRequest(msg, name);
                String query = msg.url().getRawQuery();
                yield "?" + (query != null ? query : "");
            }
            case "@query-param" -> {
                requireRequest(msg, name);
                String paramName = cid.paramString("name");
                if (paramName == null) {
                    throw new HttpSigException("@query-param requires ;name parameter");
                }
                yield extractQueryParam(msg.url(), paramName);
            }
            case "@status" -> {
                if (msg.isRequest()) {
                    throw new HttpSigException("@status only valid for response messages");
                }
                yield Integer.toString(msg.statusCode());
            }
            default -> throw new HttpSigException("unknown derived component: " + name);
        };
    }

    /**
     * Extract authority from a URI, stripping default ports (80 for http, 443 for https).
     */
    static String extractAuthority(URI uri) {
        String host = uri.getHost().toLowerCase();
        int port = uri.getPort();
        String scheme = uri.getScheme() != null ? uri.getScheme().toLowerCase() : "";
        if (port == -1
            || ("http".equals(scheme) && port == 80)
            || ("https".equals(scheme) && port == 443)) {
            return host;
        }
        return host + ":" + port;
    }

    private static String extractRequestTarget(URI uri) {
        String path = uri.getRawPath();
        if (path == null || path.isEmpty()) path = "/";
        String query = uri.getRawQuery();
        return query != null ? path + "?" + query : path;
    }

    private static String extractQueryParam(URI uri, String name) throws HttpSigException {
        String query = uri.getRawQuery();
        if (query == null) {
            throw new HttpSigException("@query-param: no query string in URI");
        }
        var values = new ArrayList<String>();
        for (String pair : query.split("&", -1)) {
            String[] kv = pair.split("=", 2);
            String decodedKey = URLDecoder.decode(kv[0], StandardCharsets.UTF_8);
            if (decodedKey.equals(name)) {
                values.add(kv.length > 1
                    ? URLDecoder.decode(kv[1], StandardCharsets.UTF_8)
                    : "");
            }
        }
        if (values.isEmpty()) {
            throw new HttpSigException("@query-param: parameter '" + name + "' not found");
        }
        // RFC 9421 says if multiple values, each one produces a separate entry.
        // For the signature base, we only handle single-value here.
        // Multiple @query-param components with the same ;name should be listed separately.
        return values.get(0);
    }

    // ---- Header fields (Section 2.1) ----

    private static String extractHeader(String name, ComponentIdentifier cid, HttpMessage msg)
            throws HttpSigException {

        List<String> values = msg.headerValues(name);

        if (cid.hasParam("bs")) {
            // Binary Sequence: each value is a separate SFV binary item,
            // combined as an inner list
            return extractByteSequence(name, values);
        }

        if (cid.hasParam("sf")) {
            // Structured Fields: re-serialize as SFV
            return extractStructuredField(name, cid, values);
        }

        if (cid.hasParam("key")) {
            // Dictionary member extraction
            return extractDictKey(name, cid, values);
        }

        // Plain header: concatenate multiple values with ", "
        if (values.isEmpty()) {
            throw new HttpSigException("header '" + name + "' not present in message");
        }
        return values.stream()
            .map(String::trim)
            .collect(Collectors.joining(", "));
    }

    private static String extractByteSequence(String name, List<String> values)
            throws HttpSigException {
        if (values.isEmpty()) {
            throw new HttpSigException("header '" + name + "' not present for ;bs");
        }
        // each value is treated as a binary item, combined in a list
        var parts = new ArrayList<String>();
        for (String v : values) {
            // the value itself is the SFV binary sequence
            parts.add(":" + java.util.Base64.getEncoder().encodeToString(
                v.trim().getBytes(StandardCharsets.UTF_8)) + ":");
        }
        return String.join(", ", parts);
    }

    private static String extractStructuredField(String name, ComponentIdentifier cid,
                                                  List<String> values) throws HttpSigException {
        if (values.isEmpty()) {
            throw new HttpSigException("header '" + name + "' not present for ;sf");
        }
        // just join and return as-is (the field is already in SFV form)
        return values.stream().map(String::trim).collect(Collectors.joining(", "));
    }

    private static String extractDictKey(String name, ComponentIdentifier cid,
                                          List<String> values) throws HttpSigException {
        if (values.isEmpty()) {
            throw new HttpSigException("header '" + name + "' not present for ;key");
        }
        String key = cid.paramString("key");
        if (key == null) {
            throw new HttpSigException(";key parameter requires a string value");
        }
        String combined = values.stream().map(String::trim).collect(Collectors.joining(", "));
        var members = SFV.parseDictionary(combined);
        for (var member : members) {
            if (member.key().equals(key)) {
                // re-serialize the member value
                if (member.value() instanceof SFV.InnerList il) {
                    return serializeInnerListForComponent(il);
                } else if (member.value() instanceof SFV.Item item) {
                    return SFV.serializeBareItem(item.value()) + serializeParams(item.params());
                }
                return SFV.serializeBareItem(member.value());
            }
        }
        throw new HttpSigException(";key member '" + key + "' not found in header '" + name + "'");
    }

    private static String serializeInnerListForComponent(SFV.InnerList il) {
        var sb = new StringBuilder();
        sb.append('(');
        for (int i = 0; i < il.items().size(); i++) {
            if (i > 0) sb.append(' ');
            sb.append(SFV.serializeBareItem(il.items().get(i).value()));
            sb.append(serializeParams(il.items().get(i).params()));
        }
        sb.append(')');
        sb.append(serializeParams(il.params()));
        return sb.toString();
    }

    private static String serializeParams(SFV.Params params) {
        if (params.isEmpty()) return "";
        var sb = new StringBuilder();
        for (var entry : params.map().entrySet()) {
            sb.append(';');
            sb.append(entry.getKey());
            Object v = entry.getValue();
            if (v instanceof Boolean b) {
                if (!b) sb.append("=?0");
            } else {
                sb.append('=');
                sb.append(SFV.serializeBareItem(v));
            }
        }
        return sb.toString();
    }

    private static void requireRequest(HttpMessage msg, String component) throws HttpSigException {
        if (!msg.isRequest()) {
            throw new HttpSigException(component + " only valid for request messages");
        }
    }
}

File: kotlin/lib/src/main/kotlin/com/zourzouvillys/httpsig/Components.kt
package com.zourzouvillys.httpsig

import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Base64

/**
 * Extracts component values from HTTP messages per RFC 9421 Section 2.
 *
 * Handles both derived components (@method, @target-uri, etc.) and
 * header fields (plain, ;sf, ;bs, ;key).
 */
internal object Components {

    /**
     * Extract the value of a component from the message.
     */
    fun extract(cid: ComponentIdentifier, msg: HttpMessage, reqMsg: HttpMessage?): String {
        // ;req means extract from the request instead
        if (cid.hasParam("req")) {
            if (reqMsg == null) {
                throw HttpSigException(";req used but no request message provided")
            }
            val strippedParams = LinkedHashMap(cid.params).apply { remove("req") }
            val strippedCid = ComponentIdentifier.withParams(cid.name, strippedParams)
            return extract(strippedCid, reqMsg, null)
        }

        return if (cid.name.startsWith("@")) {
            extractDerived(cid.name, cid, msg)
        } else {
            extractHeader(cid.name, cid, msg)
        }
    }

    // ---- Derived components (Section 2.2) ----

    private fun extractDerived(name: String, cid: ComponentIdentifier, msg: HttpMessage): String =
        when (name) {
            "@method" -> {
                requireRequest(msg, name)
                msg.method!!.uppercase()
            }
            "@target-uri" -> {
                requireRequest(msg, name)
                msg.url!!.toASCIIString()
            }
            "@authority" -> {
                requireRequest(msg, name)
                extractAuthority(msg.url!!)
            }
            "@scheme" -> {
                requireRequest(msg, name)
                msg.url!!.scheme.lowercase()
            }
            "@request-target" -> {
                requireRequest(msg, name)
                extractRequestTarget(msg.url!!)
            }
            "@path" -> {
                requireRequest(msg, name)
                val path = msg.url!!.rawPath
                if (path.isNullOrEmpty()) "/" else path
            }
            "@query" -> {
                requireRequest(msg, name)
                val query = msg.url!!.rawQuery
                "?" + (query ?: "")
            }
            "@query-param" -> {
                requireRequest(msg, name)
                val paramName = cid.paramString("name")
                    ?: throw HttpSigException("@query-param requires ;name parameter")
                extractQueryParam(msg.url!!, paramName)
            }
            "@status" -> {
                if (msg.isRequest) {
                    throw HttpSigException("@status only valid for response messages")
                }
                msg.statusCode.toString()
            }
            else -> throw HttpSigException("unknown derived component: $name")
        }

    /**
     * Extract authority from a URI, stripping default ports (80 for http, 443 for https).
     */
    internal fun extractAuthority(uri: URI): String {
        val host = uri.host.lowercase()
        val port = uri.port
        val scheme = uri.scheme?.lowercase() ?: ""
        return if (port == -1 || (scheme == "http" && port == 80) || (scheme == "https" && port == 443)) {
            host
        } else {
            "$host:$port"
        }
    }

    private fun extractRequestTarget(uri: URI): String {
        val path = uri.rawPath.let { if (it.isNullOrEmpty()) "/" else it }
        val query = uri.rawQuery
        return if (query != null) "$path?$query" else path
    }

    private fun extractQueryParam(uri: URI, name: String): String {
        val query = uri.rawQuery
            ?: throw HttpSigException("@query-param: no query string in URI")

        val values = mutableListOf<String>()
        for (pair in query.split("&")) {
            val kv = pair.split("=", limit = 2)
            val decodedKey = URLDecoder.decode(kv[0], StandardCharsets.UTF_8)
            if (decodedKey == name) {
                values.add(
                    if (kv.size > 1) URLDecoder.decode(kv[1], StandardCharsets.UTF_8) else ""
                )
            }
        }

        if (values.isEmpty()) {
            throw HttpSigException("@query-param: parameter '$name' not found")
        }
        return values[0]
    }

    // ---- Header fields (Section 2.1) ----

    private fun extractHeader(name: String, cid: ComponentIdentifier, msg: HttpMessage): String {
        val values = msg.headerValues(name)

        if (cid.hasParam("bs")) {
            return extractByteSequence(name, values)
        }

        if (cid.hasParam("sf")) {
            return extractStructuredField(name, values)
        }

        if (cid.hasParam("key")) {
            return extractDictKey(name, cid, values)
        }

        // Plain header: concatenate multiple values with ", "
        if (values.isEmpty()) {
            throw HttpSigException("header '$name' not present in message")
        }
        return values.joinToString(", ") { it.trim() }
    }

    private fun extractByteSequence(name: String, values: List<String>): String {
        if (values.isEmpty()) {
            throw HttpSigException("header '$name' not present for ;bs")
        }
        return values.joinToString(", ") { v ->
            ":" + Base64.getEncoder().encodeToString(v.trim().toByteArray(StandardCharsets.UTF_8)) + ":"
        }
    }

    private fun extractStructuredField(name: String, values: List<String>): String {
        if (values.isEmpty()) {
            throw HttpSigException("header '$name' not present for ;sf")
        }
        return values.joinToString(", ") { it.trim() }
    }

    private fun extractDictKey(name: String, cid: ComponentIdentifier, values: List<String>): String {
        if (values.isEmpty()) {
            throw HttpSigException("header '$name' not present for ;key")
        }
        val key = cid.paramString("key")
            ?: throw HttpSigException(";key parameter requires a string value")

        val combined = values.joinToString(", ") { it.trim() }
        val members = SFV.parseDictionary(combined)

        for (member in members) {
            if (member.key == key) {
                return when (val v = member.value) {
                    is SFV.InnerList -> serializeInnerListForComponent(v)
                    is SFV.Item -> SFV.serializeBareItem(v.value) + serializeParams(v.params)
                    else -> SFV.serializeBareItem(v)
                }
            }
        }
        throw HttpSigException(";key member '$key' not found in header '$name'")
    }

    private fun serializeInnerListForComponent(il: SFV.InnerList): String =
        buildString {
            append('(')
            il.items.forEachIndexed { i, item ->
                if (i > 0) append(' ')
                append(SFV.serializeBareItem(item.value))
                append(serializeParams(item.params))
            }
            append(')')
            append(serializeParams(il.params))
        }

    private fun serializeParams(params: SFV.Params): String {
        if (params.isEmpty()) return ""
        return buildString {
            for ((key, value) in params.map) {
                append(';')
                append(key)
                if (value is Boolean) {
                    if (!value) append("=?0")
                } else {
                    append('=')
                    append(SFV.serializeBareItem(value))
                }
            }
        }
    }

    private fun requireRequest(msg: HttpMessage, component: String) {
        if (!msg.isRequest) {
            throw HttpSigException("$component only valid for request messages")
        }
    }
}
